wasm/validation.cpp

```diff
@@ -0,0 +1,373 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "wasm/validation.h"
+#include "util/variant.h"
+#include "wasm/instructions.h"
+#include "wasm/types.h"
+#include "wasm/wasm.h"
+
+#include <tl/expected.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+#include <string_view>
+#include <variant>
+#include <vector>
+
+namespace wasm::validation {
+
+using namespace wasm;
+using namespace wasm::instructions;
+
+using ValueOrUnknown = std::variant<std::monostate, ValueType>;
+
+// NOLINTNEXTLINE(cert-err58-cpp)
+std::map<ValidationError, std::string_view> const validation_error_str = {
```

Comment:
Maybe replace this with a switch where each `ValidationError` case returns a string_view in `to_string` instead? No need for a nolint and no heap allocations needed with that method.